### PR TITLE
clarify system call verification data copying policies

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Expand on policies on what to do with parameters passed to system calls by reference.

Security people, please feel free to throw spitballs at this.

The special case for large data buffers I think is needed, although it will require some care in writing the implementation functions. Otherwise, there will be the overhead of both a heap allocation to duplicate the buffer and an extra copy operation. I am particularly averse to requiring a heap for what is a very common operation in syscalls.

Also, what used to be called "handler functions" are now called "verification functions", update docs for this.

### CODE REVIEW

**Comment by andrewboie** [comment]
One detail I'm not sure of: what is Zephyr's polices on VLAs? I know we have a couple verifiers that use them.

**Comment by ceolin** [comment]
I'd say we should not have VLAs, there are unspecified and undefined behaviours associated with VLAs. e.g: when the size is 0 or negative
```
The size expression in an array declaration is not a constant expression and evaluates
at program execution time to a nonpositive value (6.7.5.2). 
```
Also, it is undefined what happens when the allocation fails and there is not way to figure it out.
Another thing that Linus pointed is:
```
It generates much more code, and much _slower_ code (and more fragile code), than just using a fixed key size would have done." 
```

MISRA-C 18.8 requires to no use VLA as well

File: doc/reference/usermode/syscalls.rst [hunk 1]
```diff
@@ -332,8 +332,198 @@ For example:
...
+
+ * If the buffer is used by the implementation function to write data, such
+   as data captured from some MMIO region, the implementation function must
+   only write this data, and never read it.
+
+ * If the buffer is used by the implementation function to read data, such
+   as a block of memory to write to some hardware destination, this data
+   must be read without any processing. No conditional logic can be implemented
+   due to the data buffer's contents. If such logic is required a copy must be
+   made.
```
Comment (ceolin) [inline doc/reference/usermode/syscalls.rst:511]: you've covered my concern :)

**Review by ceolin — APPROVED** [review]
that is pretty much what we have discussed. I'm ok with that, thanks for documenting it.

**Review by dcpleung — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

File: doc/reference/usermode/syscalls.rst [hunk 2]
```diff
@@ -331,8 +332,198 @@ For example:
...
+user thread that has access to the memory that parameter points to. If the
+kernel makes any logical decisions based on the contents of this memory, this
+can open up the kernel to attacks even if checking is done. This is a class
+of exploits known as TOCTOU (Time Of Check to Time Of Use).
+
+The proper procedure to mitigate these attacks is to make a copies in the
+verification function, and only perform parameter checks on the copies, which
+user threads will never have access to. The implementation functions get passed
+the copy and not the original data sent by the user. The
+:c:func:`z_user_to_copy()` and :c:func:`z_user_from_copy()` APIs exist for
```
Comment (d3zd3z) [inline doc/reference/usermode/syscalls.rst:350]: These names are kind of confusing. Any reason to not use `z_copy_to_user` and `z_copy_from_user`?

**Review by d3zd3z — APPROVED** [review]
I think this clarification is reasonable. Things are a little ambiguous because we really haven't come up with a threat model to describe what usermode addresses. But these changes do seem reasonable things to do.

File: doc/reference/usermode/syscalls.rst (hunk 2 above)
Comment (andrewboie) [reply doc/reference/usermode/syscalls.rst:350]: these APIs already exist, and Zephyr naming convention (not followed 100%) is to have the verb last.

**Comment by andrewboie** [comment]
> Things are a little ambiguous because we really haven't come up with a thread model to describe what usermode addresses

@d3zd3z this is a serious assertion. We have a threat model posted here: https://docs.zephyrproject.org/latest/reference/usermode/index.html. **This is the first time anyone has communicated that this is insufficient**. Please provide specific, actionable feedback on what is ambiguous or missing as soon as possible. Why have you not spoken about this before??

**Comment by d3zd3z** [comment]
> We have a threat model posted here

My sincerest apologies for the unbased claim. I was not aware of the threat model.

**Pull request merged**